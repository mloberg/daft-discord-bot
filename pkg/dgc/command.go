package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
