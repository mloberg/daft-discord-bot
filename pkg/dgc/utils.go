package dgc

import "github.com/bwmarrin/discordgo"

var InteractionResponseEphemeral uint64 = 1 << 6

// GetUser returns the user that created the interaction either in a DM or guild
func GetUser(i *discordgo.InteractionCreate) *discordgo.User {
	if i.Member != nil {
		return i.Member.User
	}

	return i.User
}
