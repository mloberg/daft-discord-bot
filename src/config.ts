import joi from 'joi';
import { LevelWithSilent } from 'pino';

export const schema = joi
    .object({
        NODE_ENV: joi.string().lowercase().valid('development', 'test', 'production').default('production'),
        LOG_LEVEL: joi
            .string()
            .valid('fatal', 'error', 'warn', 'info', 'debug', 'trace', 'silent')
            .lowercase()
            .default('info'),
        DEBUG: joi.boolean().default(false),
        SONG_FILE: joi
            .string()
            .regex(/\.json$/)
            .default('songs.json'),
        BOT_PREFIX: joi.string().invalid('@').default('_'),
        BOT_TOKEN: joi.string().presence(process.env.NODE_ENV === 'test' ? 'optional' : 'required'),
    })
    .unknown()
    .required();

const { error, value: env } = schema.validate(process.env);
if (error) {
    throw new Error(`Config validation error: ${error.message}`);
}

export default {
    env: env.NODE_ENV as 'development' | 'test' | 'production',
    logLevel: env.LOG_LEVEL as LevelWithSilent,
    debug: env.DEBUG as boolean,
    dataFile: env.SONG_FILE as string,
    prefix: env.BOT_PREFIX as string,
    token: env.BOT_TOKEN as string,
};
