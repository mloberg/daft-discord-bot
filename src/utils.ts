export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
