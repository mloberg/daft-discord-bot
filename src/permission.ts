import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
