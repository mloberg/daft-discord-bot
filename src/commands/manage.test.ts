import { Client, Guild, Message, TextChannel } from 'discord.js';
import jwt from 'jsonwebtoken';
import { mocked } from 'ts-jest/utils';

import config from '../config';
import { FriendlyError } from '../error';
import { hasPermission } from '../permission';
import command from './manage';

const mocks = {
    author: jest.fn(),
    permission: mocked(hasPermission),
};

jest.mock('discord.js', () => ({
    Client: jest.fn(),
    Guild: jest.fn(),
    TextChannel: jest.fn(),
    Message: jest.fn().mockImplementation(() => ({
        member: {
            id: 'testuser',
            guild: {
                id: 'testguild',
                name: 'Test Guild',
            },
        },
        author: {
            send: mocks.author,
        },
    })),
}));

jest.mock('../permission');

describe('_manage configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('manage');
        expect(command.description).toEqual('Manage songs via web interface');
        expect(command.usage).toBeUndefined();
    });

    it('should have a web alias', () => {
        expect(command.alias).toContain('web');
    });
});

describe('_manage', () => {
    let message: Message;

    beforeEach(() => {
        mocks.author.mockClear();
        mocks.permission.mockClear();

        const client = new Client();
        const guild = new Guild(client, {});
        const channel = new TextChannel(guild, {});
        message = new Message(client, {}, channel);
    });

    it('adds a file to the manager with a title', async () => {
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: [], $0: 'manage' });

        expect(mocks.author).toHaveBeenCalledTimes(1);

        const msg = mocks.author.mock.calls[0][0];
        const match = msg.match(/^Manage songs at http:\/\/localhost:53134\/\?token=(.+)/);
        const data = jwt.verify(match[1], config.secret);

        expect(data).toMatchObject({ guild: 'testguild', name: 'Test Guild', member: 'testuser' });
    });

    it('will throw an error if user does not have role', async () => {
        mocks.permission.mockReturnValue(false);

        try {
            await command.run(message, { _: [], $0: 'web' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('You do not have permission to do that.');
        }
    });
});
