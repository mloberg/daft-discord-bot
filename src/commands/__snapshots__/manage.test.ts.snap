// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_manage is a command 1`] = `
Object {
  "alias": Array [
    "web",
  ],
  "description": "Manage songs via web interface",
  "name": "manage",
  "run": [Function],
}
`;
